### ISSUE TITLE
ActionController::Parameters#== does not works as expected

### ISSUE THREAD

**Issue by meinac** [rails/rails#21032]
`ActionController::Parameters#==` will return false even if the hash and AC::Parameters hash same because of the `HashWithIndifferentAccess`. But the documentation of `==` method says we can compare `ActionController::Parameters` with `Hash-Like` objects.
```ruby
hash = { foo: :bar }
params = ActionController::Parameters.new(hash)
params == hash #returns false
```
I believe this is not the bug about `ActionController::Parameters` this is about `HashWithIndifferentAccess`. I think we should implement `==`, `eql?` methods in `HashWithIndifferentAccess` to give correct results. If you are OK with implementing these methods in `HashWithIndifferentAccess` I'm going to finish my initial work on it.

**Comment by rafaelfranca** [comment]
:+1:

**Comment by atul-shimpi** [comment]
Yes,
HashWithIndifferent will convert symbol keys to string and comparing string and symbol will return false.
hash = { foo: :bar } will get converted to Params {"foo"=>:bar} and  foo: == "foo" wil return false.

**Comment by maclover7** [comment]
Closed via #23167.
